File: Android/PlayerProj/animplayer/src/main/java/com/tencent/qgame/animplayer/mix/Src.kt
/*
 * Tencent is pleased to support the open source community by making vap available.
 *
 * Copyright (C) 2020 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * Licensed under the MIT License (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tencent.qgame.animplayer.mix

import android.graphics.Bitmap
import android.graphics.Color
import com.tencent.qgame.animplayer.Constant
import com.tencent.qgame.animplayer.util.ALog
import org.json.JSONObject

class Src {
    companion object {
        private const val TAG = "${Constant.TAG}.Src"
    }

    enum class SrcType(val type: String) {
        UNKNOWN("unknown"),
        IMG("img"),
        TXT("txt"),
    }

    enum class LoadType(val type: String) {
        UNKNOWN("unknown"),
        NET("net"), // 网络加载的图片
        LOCAL("local"), // 本地加载的图片
    }

    enum class FitType(val type: String) {
        FIT_XY("fitXY"), // 按原始大小填充纹理
        CENTER_FULL("centerFull"), // 以纹理中心点放置
    }

    enum class Style(val style: String) {
        DEFAULT("default"),
        BOLD("b"), // 文字粗体
    }

    var srcId = ""
    var w = 0
    var h = 0
    var drawWidth = 0
    var drawHeight = 0
    var srcType = SrcType.UNKNOWN
    var loadType = LoadType.UNKNOWN
    var srcTag = ""
    var txt = ""
    var style = Style.DEFAULT
    var color: Int = 0
    var fitType = FitType.FIT_XY
    var srcTextureId = 0
    var bitmap: Bitmap? = null
        set(value) {
            field = value
            genDrawSize(value)
        }

    constructor(json: JSONObject) {
        srcId = json.getString("srcId")
        w = json.getInt("w")
        h = json.getInt("h")
        // 可选
        var colorStr = json.optString("color", "#000000")
        if (colorStr.isEmpty()) {
            colorStr = "#000000"
        }
        color = Color.parseColor(colorStr)
        srcTag = json.getString("srcTag")
        txt = srcTag

        srcType = when(json.getString("srcType")) {
            SrcType.IMG.type -> SrcType.IMG
            SrcType.TXT.type -> SrcType.TXT
            else -> SrcType.UNKNOWN
        }
        loadType = when(json.getString("loadType")) {
            LoadType.NET.type -> LoadType.NET
            LoadType.LOCAL.type -> LoadType.LOCAL
            else -> LoadType.UNKNOWN
        }
        fitType = when(json.getString("fitType")) {
            FitType.CENTER_FULL.type -> FitType.CENTER_FULL
            else -> FitType.FIT_XY
        }

        // 可选
        style = when(json.optString("style", "")) {
            Style.BOLD.style -> Style.BOLD
            else -> Style.DEFAULT
        }
        ALog.i(TAG, "${toString()} color=$colorStr")
    }


    private fun genDrawSize(bitmap: Bitmap?) {
        val bw = bitmap?.width?: w
        val bh = bitmap?.height?: h
        drawWidth = bw
        drawHeight = bh
        if (fitType == FitType.CENTER_FULL) {
            if (w == 0 || h == 0) {
                return
            }
            // 按src w h进行centerCrop处理
            val srcRate = w.toFloat() / h.toFloat()
            val bitmapRate = bw.toFloat() / bh.toFloat()

            if (bitmapRate >= srcRate) {
                drawHeight = h
                drawWidth = (h * bitmapRate).toInt()
            } else {
                drawWidth = w
                drawHeight = (w / bitmapRate).toInt()
            }
        }
    }


    override fun toString(): String {
        return "Src(srcId='$srcId', srcType=$srcType, loadType=$loadType, srcTag='$srcTag', bitmap=$bitmap, txt='$txt')"
    }

}


class SrcMap(json: JSONObject) {
    val map = HashMap<String, Src>()

    init {
        val srcJsonArray = json.getJSONArray("src")
        val srcLen = srcJsonArray?.length() ?: 0
        for (i in 0 until srcLen) {
            val srcJson = srcJsonArray?.getJSONObject(i) ?: continue
            val src = Src(srcJson)
            if (src.srcType != Src.SrcType.UNKNOWN) { // 不认识的srcType丢弃
                map[src.srcId] = src
            }
        }
    }
}

File: ohos/vap_module/src/main/cpp/mix/src.cpp
/*
 * Copyright (C) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "src.h"
#include "log.h"
#include "common_const.h"

using namespace CommonConst;

std::map<std::string, SrcType> Src::string2SrcType = {
    {"unknown", SrcType::UNKNOWN}, {"img", SrcType::IMG}, {"txt", SrcType::TXT}};

std::map<std::string, LoadType> Src::string2LoadType = {
    {"unknown", LoadType::UNKNOWN}, {"net", LoadType::NET}, {"local", LoadType::LOCAL}};

std::map<std::string, FitType> Src::string2FitType = {{"fitXY", FitType::FIT_XY}, {"centerFull", FitType::CENTER_FULL}};

std::map<std::string, Style> Src::string2Style = {{"default", Style::DEFAULT}, {"b", Style::BOLD}};

Src::Src(json jsonSrc)
{
    srcId = jsonSrc.at("srcId").get<std::string>();
    std::string tmpStr = jsonSrc.at("srcType").get<std::string>();
    srcType = stringToSrcTypeFunc(tmpStr);
    tmpStr = jsonSrc.at("loadType").get<std::string>();
    loadType = stringToLoadTypeFunc(tmpStr);
    srcTag = jsonSrc.at("srcTag").get<std::string>();
    txt = srcTag;
    std::string colorStr;
    if (jsonSrc.find("color") != jsonSrc.end()) {
        colorStr = jsonSrc.at("color").get<std::string>();
        colorStr.erase(0, 1); // 去掉字符串开头的 #
        color = std::stoi(colorStr, nullptr, TWO_TIME_EIGHT);
        if (((color >> THREE_TIME_EIGHT) & 0xff) == 0) {
            color |= 0xff << THREE_TIME_EIGHT;
        }
    }
    if (jsonSrc.find("style") != jsonSrc.end()) {
        tmpStr = jsonSrc.at("style").get<std::string>();
        style = stringToStyleFunc(tmpStr);
    }
    w = jsonSrc.at("w").get<int32_t>();
    h = jsonSrc.at("h").get<int32_t>();
    tmpStr = jsonSrc.at("fitType").get<std::string>();
    fitType = stringToFitTypeFunc(tmpStr);
}

void Src::PrintInfo()
{
    LOGD("srcId: %{public}s srcType: %{public}d, loadType:%{public}d, srcTag:%{public}s,"
        "color:%{public}x, style:%{public}d, w:%{public}d, h:%{public}d, fitType: %{public}d, "
        "drawWidth: %{public}d,drawHeight: %{public}d, srcTextureId: %{public}d",
        srcId.c_str(), srcType, loadType, srcTag.c_str(), color, style, w, h, fitType, drawWidth, drawHeight,
        srcTextureId);
}

SrcMap::SrcMap(json jsonObj)
{
    auto srcJsonArray = jsonObj.at("src");
    size_t srcLen = srcJsonArray.size();
    for (int i = 0; i < srcLen; i++) {
        auto srcJson = srcJsonArray[i];
        Src src = Src(srcJson);
        if (src.srcType != SrcType::UNKNOWN) { // 不认识的srcType丢弃
            srcSMap.insert(std::make_pair(src.srcId, src));
        }
    }
}